src/main/java/com/thealgorithms/sorts/SmoothSort.java

```diff
@@ -0,0 +1,227 @@
+                if (prevOneIndex != -1) {
+                    answer[0] = prevOneIndex;
+                    answer[1] = i;
+                }
+                prevOneIndex = i;
+            } else {
+                prevOneIndex = -1;
+            }
+            num >>>= 1;
+        }
+        return answer;
+    }
+
+    private static Integer[] findAllLeonardoTreeIndices(int num) {
+        int setBitCount = 0;
+        for (int i = 0; i < Integer.SIZE; i++) {
+            if ((num & (1 << i)) != 0) {
+                setBitCount++;
+            }
+        }
+
+        Integer[] setBitIndexes = new Integer[setBitCount];
+        int index = 0;
+        for (int i = 0; i < Integer.SIZE; i++) {
+            if ((num & (1 << i)) != 0) {
+                setBitIndexes[index++] = i;
+            }
+        }
+        return setBitIndexes;
+    }
+
+    private static <T extends Comparable<T>> void shiftRootAndRestoreHeap(int lenardoLevelTracker, int leonardoHeapSize, T[] array) {
+
+        if (leonardoHeapSize == 0) {
+            return;
+        }
+
+        Integer[] currentLeonardoTreeLevels = findAllLeonardoTreeIndices(lenardoLevelTracker);
+        int previousTreeSizeCumulative = 0;
+        ArrayList<Integer> rootNodeIndices = new ArrayList<Integer>();
+        Collections.reverse(Arrays.asList(currentLeonardoTreeLevels)); // To get the Levels in decreasing order of levels
+
+        // The number of roots are going to be same the the number of levels
+        // iterate over the currentLeonardoTreeLevels and get roots
+
+        for (int i = 0; i < currentLeonardoTreeLevels.length; i++) {
+            rootNodeIndices.add(previousTreeSizeCumulative + getLeonardoNumbers()[currentLeonardoTreeLevels[i]] - 1);
+            previousTreeSizeCumulative = previousTreeSizeCumulative + getLeonardoNumbers()[currentLeonardoTreeLevels[i]];
+        }
+
+        int rootNodeIndexForHeapify = rootNodeIndices.getLast();
+        int leonardoTreeLevelforHeapify = currentLeonardoTreeLevels[currentLeonardoTreeLevels.length - 1];
+
+        for (int i = 0; i < rootNodeIndices.size(); i++) {
+            if (i == 0) {
+                continue;
+            }
+
+            int currentRootNodeIndex = rootNodeIndices.get(i);
+            int prevRootNodeIndex = rootNodeIndices.get(i - 1);
+            int j = i;
+            while (array[prevRootNodeIndex].compareTo(array[currentRootNodeIndex]) > 0) {
+                int currentLeonardoLevel = currentLeonardoTreeLevels[j];
+                if (currentLeonardoLevel > 1) {
+                    // compare child and swap
+
+                    int indexOfRightChild = rootNodeIndices.get(j) - 1; // right child is of level n-2
+                    int indexOfLeftChild = rootNodeIndices.get(j) - 1 - getLeonardoNumbers()[currentLeonardoLevel - 2];
+                    if (array[prevRootNodeIndex].compareTo(array[indexOfRightChild]) > 0 && array[prevRootNodeIndex].compareTo(array[indexOfLeftChild]) > 0) {
```

Comment:
Sure, I shall add a specific test case in `SmoothSortTest.java`.

Reply:
I see why this is happening and I shall try my best to explain in this comment. 

Leonardo Heap has the following three properties that define it (Going through the array from left to right):
1. The trees are in decreasing order of levels
2. Root nodes of the trees are in increasing order
3. Each tree is in a 'max-heap' state. 

The Level-1 leonardo Tree(L1) and Level 0 leonardo Tree (L0) are the 'base case' trees as they have exactly one node. 

At any given valid leonardo heap, if the heap contains both L1 and L0 tree then the element in L0 tree will always be greater than element at L1 tree, because of the second property mentioned above ( L0 >= L1 )

When L2 tree is being constructed, the level0 tree becomes the right child level1 tree becomes the left child. And hence it is not possible to have a condition such that level0 < prevRootNode < level1 as it would be a voilation of the above. 

[This](https://en.wikipedia.org/wiki/Smoothsort#Sifting_down) section of the wiki page and [this ](https://www.keithschwarz.com/smoothsort/) article (under the image`An erroneous swap` ) mention that root needs to be bigger than the new element and the roots of its child nodes for a swap to happen. However [this ](https://www.enterag.ch/hartwig/order/smoothsort.pdf) article(in page 6) mentions that the new elment should exceede its largest son. 

I believe the statement in wiki is more genreic statement than a technical one and this condition can be optimized to just having `array[prevRootNodeIndex].compareTo(array[indexOfLeftChild]) > 0`. I shall however look around for more articles that suggest otherwise before I make changes to this.